### PR TITLE
core(gather-runner): error on non-HTML

### PR DESCRIPTION
**Pull request by lemcardenas**
**Summary**

This PR adds a function in gather-runner.js, getNonHtmlError, that brings up an error if we attempt to run gatherers on non-HTML webpages.

**Related Issues/PRs**

https://github.com/GoogleChrome/lighthouse/issues/9245

### CODE REVIEW

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -215,6 +215,31 @@ class GatherRunner {
...
   }
 
+  /**
+   * Returns an error if we try to load a non-HTML page.
+   * @param {LH.Artifacts.NetworkRequest|undefined} mainRecord
+   * @return {LH.LighthouseError|undefined}
+   */
+  static getDocTypeError(mainRecord) {
+    // MIME types are case-insenstive
+    const HTML_MIME_REGEX = /^text\/html$/i;
```
Comment (lemcardenas) [inline lighthouse-core/gather/gather-runner.js:225]: According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types, "MIME types are case-insensitive but are traditionally written in lowercase". I am unsure if the check needs to be this rigorous/if a constant 'text/html' is good enough.

File: lighthouse-core/gather/gather-runner.js [hunk 2]
```diff
@@ -246,6 +272,9 @@ class GatherRunner {
     // Example: `DNS_FAILURE` is better than `NO_FCP`.
     if (networkError) return networkError;
 
+    // We want to error when the page is not of MIME type text/html
+    if (docTypeError) return docTypeError;
```
Comment (lemcardenas) [inline lighthouse-core/gather/gather-runner.js:276]: I assumed that the order of importance of errors was interstitial -> network -> docType, should this be the case?

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 3]
```diff
@@ -1095,6 +1096,52 @@ describe('GatherRunner', function() {
...
+      const mainRecord = new NetworkRequest();
+      const mimeType = 'text/html';
+      mainRecord.url = url;
+      mainRecord.mimeType = mimeType;
+      expect(GatherRunner.getDocTypeError(mainRecord)).toBeUndefined();
+    });
+
+    it('fails when the page is not of MIME type text/html', () => {
+      const url = 'http://the-page.com';
+      const mimeType = 'application/xml';
```
Comment (lemcardenas) [inline lighthouse-core/test/gather/gather-runner-test.js:1133]: For this test & unit testing in general, should I apply a rigorous run through of incorrect mimeTypes, or is just having one incorrect option good enough?

File: lighthouse-core/lib/lh-error.js [hunk 4]
```diff
@@ -47,6 +47,8 @@ const UIStrings = {
   internalChromeError: 'An internal Chrome error occurred. Please restart Chrome and try re-running Lighthouse.',
   /** Error message explaining that fetching the resources of the webpage has taken longer than the maximum time. */
   requestContentTimeout: 'Fetching resource content has exceeded the allotted time',
+  /** Error message explaining that the webpage is non-HTML, so audits are ill-defined **/
+  docTypeInvalid: 'The webpage you have provided appears to be non-HTML',
```
Comment (connorjclark) [inline lighthouse-core/lib/lh-error.js:51]:
```suggestion
  docTypeInvalid: 'The page provided is not HTML',
```

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 5]
```diff
@@ -1114,26 +1161,30 @@ describe('GatherRunner', function() {
       navigationError = /** @type {LH.LighthouseError} */ (new Error('NAVIGATION_ERROR'));
     });
 
-    it('passes when the page is loaded', () => {
+    it('passes when the page is loaded and doc type is text/html', () => {
```
Comment (connorjclark) [inline lighthouse-core/test/gather/gather-runner-test.js:1164]: for these tests, I think you can 

1) leave the test name the same
2) prefer `mainRecord.mimeType = 'text/html'` over defining a variable

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 6]
```diff
@@ -1193,7 +1260,9 @@ describe('GatherRunner', function() {
       const mainRecord = new NetworkRequest();
       const loadData = {networkRecords: [mainRecord]};
 
+      const mimeType = 'text/html';
       mainRecord.url = passContext.url;
+      mainRecord.mimeType = mimeType;
```
Comment (connorjclark) [inline lighthouse-core/test/gather/gather-runner-test.js:1265]: same here: inline the mimeType

File: lighthouse-core/gather/gather-runner.js [hunk 7]
```diff
@@ -246,6 +272,9 @@ class GatherRunner {
     // Example: `DNS_FAILURE` is better than `NO_FCP`.
     if (networkError) return networkError;
 
+    // We want to error when the page is not of MIME type text/html
```
Comment (connorjclark) [inline lighthouse-core/gather/gather-runner.js:275]:
```suggestion
    // Error if page is not HTML.
```

File: lighthouse-core/gather/gather-runner.js [hunk 8]
```diff
@@ -233,6 +258,7 @@ class GatherRunner {
 
     const networkError = GatherRunner.getNetworkError(mainRecord);
     const interstitialError = GatherRunner.getInterstitialError(mainRecord, networkRecords);
+    const docTypeError = GatherRunner.getDocTypeError(mainRecord);
```
Comment (connorjclark) [inline lighthouse-core/gather/gather-runner.js:261]: The structure of this function is rather strange–zooming out a bit, I'd expect the `loadFailureMode` check to happen first; and for each of these errors to be created in order, but only if the former error check didn't hit. Incrementally, what you have here makes sense, I just think this entire function might benefit from a slight refactor.

File: lighthouse-core/gather/gather-runner.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js:225]: would `mainRecord.resourceType === NetworkRequest.RESOURCE_TYPES.Document` work too? perhaps not for xml docs...

Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:225]: just double checked and xml docs would indeed pass (no error) if we used that condition

File: lighthouse-core/gather/gather-runner.js (hunk 7 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js]: Thanks, fixed this

File: lighthouse-core/lib/lh-error.js (hunk 4 above)
Comment (lemcardenas) [reply lighthouse-core/lib/lh-error.js]: Thanks, fixed this

File: lighthouse-core/test/gather/gather-runner-test.js (hunk 5 above)
Comment (lemcardenas) [reply lighthouse-core/test/gather/gather-runner-test.js]: Thanks, fixed both of these

File: lighthouse-core/gather/gather-runner.js (hunk 8 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:261]: Yeah, there is also definitely some redundant checks in each get function that don't need to be there if it was refactored. Should I try to refactor in this PR? It looks like Patrick wrote this function originally, I can also ping them about this if required.

Comment (patrickhulce) [reply lighthouse-core/gather/gather-runner.js:261]: @connorjclark is the refactoring we're talking about here moving 1 line a few lines up and 1 more line a few lines down? I think we can manage it here 😉 

be my guest! I certainly have no strong attachment to the specific order in which they're declared as long as we don't let them become multi-line statements :)

File: lighthouse-core/gather/gather-runner.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gather-runner.js:225]: does the mime type in the network records return Chrome's inferred mime type or just what was set on the header?

we should test it, but since the failure mode is so drastic, I would prefer to keep the no mimeType case passing (i.e. only reject documents that have an explicit non-HTML mimeType set)

File: lighthouse-core/gather/gather-runner.js [hunk 9]
```diff
@@ -215,6 +215,31 @@ class GatherRunner {
...
+   */
+  static getDocTypeError(mainRecord) {
+    // MIME types are case-insenstive
+    const HTML_MIME_REGEX = /^text\/html$/i;
+
+    // If we never requested a document, there's no doctype error, let other cases handle it.
+    if (!mainRecord) return undefined;
+
+    // If the main document failed, this error case is undefined, let other cases handle it.
+    if (mainRecord.failed) return undefined;
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js:231]: probably fine to combine these two cases into a single `if (!mainRecord || mainRecord.failed) return;` since they aren't super important to this check in particular, but could probably just drop the `mainRecord.failed` check completely, since this function isn't so much a check of a valid main document than a check that it's not an invalid media type.

File: lighthouse-core/gather/gather-runner.js (hunk 8 above)
Comment (brendankenny) [reply lighthouse-core/gather/gather-runner.js:261]: > moving 1 line a few lines up and 1 more line a few lines down?

yeah, that sounds like no problem, but I would be hesitant to do much deduplicating of content within the error checks. They work well today and it's easy to reason about them since they're more or less independent of each other, each just checking on their own particular error case.

Maybe pulling out the `!mainRecord` check makes sense since they all end up having to check that and the type system will keep track of it already being done (since we can drop all the `|undefined`s)? But, again, everything's working today and this function isn't in bad shape or anything, so @lemcardenas you can also feel free to leave it to a follow-up PR or to someone else in the future if it feels like the PR scope is growing too much.

https://google.github.io/eng-practices/review/developer/small-cls.html is one of my favorite docs :)

File: lighthouse-core/test/gather/gather-runner-test.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/test/gather/gather-runner-test.js:1133]: > For this test & unit testing in general, should I apply a rigorous run through of incorrect mimeTypes, or is just having one incorrect option good enough?

It depends on the situation. The mimeType string part of the check is a relatively simple comparison, it's basically just string equality with no real corner cases, so this seems fine. Since it's going to be case insensitive, maybe add a test for that case?

**Comment by brendankenny** [comment]
> does the mime type in the network records return Chrome's inferred mime type or just what was set on the header?

Looks like maybe it's just what was on the header, or maybe Chrome does some inference on top of that?

Maybe we'll need to combine `mimeType` with other information? It looks like checking for `'text/html'` would catch a large majority of cases but would miss some. Since this will be a fatal error, we'll probably want to be more conservative in throwing. e.g. a server shouldn't be serving html with mime type `'content/html'`, but if Lighthouse can still run against it, we shouldn't not let them do so.

This is a query of the June 2020 HTTPArchive run for unique main document mimeTypes and an example requested/finalUrl for each. Out of 5.7 million LH runs (results with no network requests or fatal load errors eliminated):

(note that any of these URLs could be NSFW)

| mimeType                    | count   | example requestedUrl              | finalUrl                                                                                                                                                                                                              | 
|-----------------------------|---------|--------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| 
| text/html                   | 5628537 | `http://0-1.ru/`                             | `http://0-1.ru/`                                                                                                                                                                                                        | 
| (_empty string_)              | 32019   | `http://100.v523.com.tw/`                    | `http://100.v523.com.tw/estate/index.do`                                                                                                                                                                                | 
| text/plain                  | 5043    | `http://52kartu.com/`                        | `http://52kartu.com/mobile.php;jsessionid=9642EBFAF295C89B0EC9AFC264AA2E63?0`                                                                                                                                           | 
| httpd/unix-directory        | 164     | `http://code.sfu.ca/`                        | `http://code.sfu.ca/index.html`                                                                                                                                                                                         | 
| application/xhtml+xml       | 115     | `http://apple-store.in/`                     | `http://apple-store.in/`                                                                                                                                                                                                | 
| text/xml                    | 93      | `http://alertas.myweb.vodafone.pt/`          | `http://alertas.myweb.vodafone.pt/home`                                                                                                                                                                                 | 
| application/octet-stream    | 85      | `http://ac.dragonest.com/`                   | `http://ac.dragonest.com/m/en`                                                                                                                                                                                          | 
| application/json            | 80      | `http://asistencia.claro.com.co/`            | `http://asistencia.claro.com.co/soporte/`                                                                                                                                                                               | 
| text/x-perl                 | 31      | `http://azerbaijan.mfa.gov.by/`              | `http://azerbaijan.mfa.gov.by/ru/`                                                                                                                                                                                      | 
| (_null_)                      | 25      | `http://aplicacionesbiblioteca.udea.edu.co/` | about:blank                                                                                                                                                                                                           | 
| application/binary          | 24      | `http://ads.google.com/`                     | `https://ads.google.com/home/`                                                                                                                                                                                          | 
| application/x-httpd-cgi     | 17      | `http://post.japanpost.jp/`                  | `https://www.post.japanpost.jp/index.html`                                                                                                                                                                              | 
| application/cgi             | 12      | `http://filedwon.info/`                      | `http://filedwon.info/?op=login`                                                                                                                                                                                        | 
| text/x-unknown-content-type | 10      | `http://gov.garant.ru/`                      | `http://gov.garant.ru/SESSION/PDA/main.htm`                                                                                                                                                                             | 
| application/x-gzip          | 8       | `http://www.wise.org/`                       | `http://www.wise.org/en_US/index.html`                                                                                                                                                                                  | 
| application/pdf             | 7       | `https://www.corelaboratory.abbott/`         | `https://www.corelaboratory.abbott/us/en/home`                                                                                                                                                                          | 
| application/x-msdownload    | 6       | `http://www.actorstudio.fr/`                 | `http://www.actorstudio.fr/FR/`                                                                                                                                                                                         | 
| text/x-python               | 5       | `http://www.dramonline.org/`                 | `http://www.dramonline.org/`                                                                                                                                                                                            | 
| application/xml             | 4       | `http://www.insan-pratama.com/`              | `http://www.insan-pratama.com/home.html`                                                                                                                                                                                | 
| application/x-httpd-fcgi    | 3       | `http://automower-fans.les-forums.com/`      | `http://automower-fans.les-forums.com/forums/`                                                                                                                                                                          | 
| application/x-perl          | 3       | `https://pedidoscastalia.com/`               | `https://pedidoscastalia.com/_login`                                                                                                                                                                                    | 
| text/javascript             | 3       | `http://olink.tv/`                           | `https://whos.amung.us/pingjs/?k=djmatioka403&amp;t=216.244.91.194__&c=d&y=Y20200614164115&a=0&r=20200614164115`                                                                                                        | 
| text/vnd.wap.wml            | 2       | `http://m.n-content.com/`                    | `http://m.n-content.com/MusicMobile/wap/index.jsp?cid=1&ads=0007`                                                                                                                                                       | 
| application/unknown         | 2       | `https://cvd.bundesregierung.de/`            | `https://m.cvd.bundesregierung.de/cvdm-de/login`                                                                                                                                                                        | 
| application/javascript      | 2       | `http://map.baidu.com/`                      | `https://map.baidu.com/mobile/webapp/index/index/`                                                                                                                                                                      | 
| text/x-server-parsed-html   | 2       | `http://act.zisho.jp/`                       | `http://act.zisho.jp/`                                                                                                                                                                                                  | 
| "text/html"                 | 1       | `https://www.svobodny-statek.cz/`            | `https://www.svobodny-statek.cz/bio-dynamicke-zemedelstvi`                                                                                                                                                              | 
| application/x-cgi           | 1       | `https://www.jointheblue.com/`               | `https://www.jointheblue.com/michigan/`                                                                                                                                                                                 | 
| application/x-javascript    | 1       | `http://ichinoseki-shinkin.jp/`              | `http://ichinoseki-shinkin.jp/sp/index.html`                                                                                                                                                                            | 
| image/jpeg                  | 1       | `http://www.oeuvrecoeurcroix.fr/`            | `http://www.oeuvrecoeurcroix.fr/portail/`                                                                                                                                                                               | 
| audio/mpeg                  | 1       | `https://www.radiostella.cz/`                | `https://ice3.abradio.cz/stella128.mp3`                                                                                                                                                                                 | 
| image/gif                   | 1       | `http://ww3.speetest.net/`                   | `https://partners.etoro.com/blank.gif`                                                                                                                                                                                  | 
| text/html/html              | 1       | `https://access.sheridaninstitute.ca/`       | `https://access.sheridaninstitute.ca/http://portal-am.sheridaninstitute.ca/amserver/UI/Login?gw=access.sheridaninstitute.ca&org=o%3Dsheridanc.on.ca&goto=http%3A%2F%2Fportal-p.sheridaninstitute.ca%3A80%2Fportal%2Fdt` | 
| content/html                | 1       | `https://subs.nymag.com/`                    | `https://subs.nymag.com/account/`                                                                                                                                                                                       |

**Comment by brendankenny** [comment]
Actually, some of these are wrong and are the mimeType of the `requestedUrl` even if there's a redirect to the `finalUrl` (it's kind of hard to find the main document because AFAICT there's no URL constructor available in the BigQuery custom JS functions), but I'm leaving it up because they appear to all be real initial mimeTypes

e.g. `https://www.svobodny-statek.cz/` does have mimeType `"text/html"` (in double quotes), even though the URL it's redirected to has the correct mimeType `text/html` (same with the first request of `http://www.oeuvrecoeurcroix.fr/` being served as `image/jpeg`).

I can post the better table but it's mostly just a subset of the above, with far more cases correctly under `text/html` (5,665,445 `text/html`, 605 empty strings, 126 `application/xhtml+xml`, 88 `text/plain`, 7 `text/xml`, and then various other ones all occurring less than 3 times).

Regardless, looks like this string can be whatever the server wants it to be.

File: lighthouse-core/gather/gather-runner.js [hunk 10]
```diff
@@ -215,6 +215,31 @@ class GatherRunner {
     return new LHError(LHError.errors.CHROME_INTERSTITIAL_ERROR);
   }
 
+  /**
+   * Returns an error if we try to load a non-HTML page.
+   * @param {LH.Artifacts.NetworkRequest|undefined} mainRecord
+   * @return {LH.LighthouseError|undefined}
+   */
+  static getDocTypeError(mainRecord) {
```
[the remark this answers is not in the document]
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:223]: Sounds good, I wondered why docType/doctype sounded so familiar to me. I can definitely change the name to what you recommended

File: lighthouse-core/gather/gather-runner.js (hunk 8 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:261]: I think I'll try to stay safe for this PR and won't change anything about the getPageLoadError function, because as mentioned below there may be more work needed for the original change to work.

**Comment by paulirish** [comment]
nice query!

> Maybe we'll need to combine `mimeType` with other information? It looks like checking for `'text/html'` would catch a large majority of cases but would miss some. Since this will be a fatal error, we'll probably want to be more conservative in throwing. e.g. a server shouldn't be serving html with mime type `'content/html'`, but if Lighthouse can still run against it, we shouldn't not let them do so.

ehhhh. im happy to yell at `content/html` even tho technically the page will render.

Because we have the browser sniff, it complicates things. my assumption is that the `mimeType` from the protocol is post-sniff but i'm not confident.

broad categories we have

1. page is served with text/html and is text/html. YAY
1. page is served with non-text/html and isn't HTML. page (probably) renders fine?
   - we throw error.
1. page is served with non-text/html but is HTML. page renders fine.
   - technically the page is good enough for us, but we should throw this error anyway. 
1. page is not served with a content-type, but is HTML and the browser sniffs it works. YAY
1. page is not served with a content-type, but isnt HTML and the browser sniffs and who knows.
   - Tough situation. LH should throw on this. I have no idea what the `.mimeType` is in this scenario, but from the above comment it sounds like it's _not_ text/html.


---------


given that we're talking about 0.015% of pages in HA.. I think we can afford to just flag everything that's without a .mimeType of `text/html`.

**Comment by brendankenny** [comment]
> page is served with non-text/html but is HTML. page renders fine.
> technically the page is good enough for us, but we should throw this error anyway.

I can kind of get behind this, but this will be very fatal (`lhrRuntimeError: true`), so some number of currently OK users won't be able to run Lighthouse until they reconfigure their server. It is a very small number, true, though how representative the HTTPArchive mime type numbers are of the general population, I don't know.

File: lighthouse-core/lib/lh-error.js [hunk 11]
```diff
@@ -312,6 +314,11 @@ const ERRORS = {
     message: UIStrings.pageLoadFailedHung,
     lhrRuntimeError: true,
   },
+  /* Used when the page is non-HTML. */
+  INVALID_DOC_TYPE: {
+    code: 'INVALID_DOC_TYPE',
+    message: UIStrings.docTypeInvalid,
```
Comment (brendankenny) [inline lighthouse-core/lib/lh-error.js:320]: need to also add `lhrRuntimeError: true` here because we want the error to bubble all the way to the top and show up in the LHR [`runtimeError` entry](https://github.com/GoogleChrome/lighthouse/blob/d9b3caba902a54420b3cab4c128b69e0e0a115c0/types/lhr.d.ts#L56) and exit with a non-zero exit code when running the CLI.

@connorjclark's examples in #9245 show that things *mostly* go wrong but not completely, so we need to be extra loud to get the attention of users doing automated runs of LH that they're accidentally auditing a page they either didn't mean to or have misconfigured.

Since it'll be `lhrRuntimeError: true` (and so can also appear in the LHR that PSI serves), it'll also need to be added as an [error entry in the proto](https://github.com/GoogleChrome/lighthouse/blob/d9b3caba902a54420b3cab4c128b69e0e0a115c0/proto/lighthouse-result.proto#L12).

File: lighthouse-core/lib/lh-error.js [hunk 12]
```diff
@@ -47,6 +47,8 @@ const UIStrings = {
   internalChromeError: 'An internal Chrome error occurred. Please restart Chrome and try re-running Lighthouse.',
   /** Error message explaining that fetching the resources of the webpage has taken longer than the maximum time. */
   requestContentTimeout: 'Fetching resource content has exceeded the allotted time',
+  /** Error message explaining that the webpage is non-HTML, so audits are ill-defined **/
+  docTypeInvalid: 'The page provided is not HTML',
```
Comment (brendankenny) [inline lighthouse-core/lib/lh-error.js:51]: maybe we should parameterize this message, so the user knows what to fix? Something like
```suggestion
  docTypeInvalid: 'The page provided is not HTML (served as MIME type {mimeType}).',
```

**Comment by lemcardenas** [comment]
> Regardless, looks like this string can be whatever the server wants it to be.

Do you think that looking directly at Content-Type headers will help in reducing some of the false negatives / be better overall? When I first asked about choosing between mimeType/Content-Type/extension approaches, Paul linked this [code](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/inspector/inspector_network_agent.cc;l=628;drc=eae84e4d756317f440495bf284f1bcb66247bdcf) which mentions `//Use mime type from cached resource in case the one in response is empty.`, when checking some of the calling [functions](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/platform/loader/fetch/resource_response.cc;drc=f5c26dea72747aa434da84c57655cac5272d5ad4;l=142) it mentions `//MIME type is determined by HTTP Content-Type header`. So from these snippets it sounds like Content-Type may be the stronger method.

In the original issue for this PR, Connor mentioned that 

> Images, stylesheets, js, and such actually audit just fine (rather, the page Chrome builds to view the resource presents no issues for LH), so we don't have to exclude those.

Should we instead then have deny/allow lists to have more coverage?

File: lighthouse-core/lib/lh-error.js (hunk 11 above)
Comment (lemcardenas) [reply lighthouse-core/lib/lh-error.js:320]: Thanks, I'll make sure to fix this!

File: lighthouse-core/lib/lh-error.js (hunk 12 above)
Comment (lemcardenas) [reply lighthouse-core/lib/lh-error.js:51]: I like that idea, my initial message had the mimeType information but wasn't as easy to read as this, I'll change it!

File: lighthouse-core/gather/gather-runner.js (hunk 9 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:231]: Thanks for pointing this out, I ended up just dropping the mainRecord.failed check

**Comment by paulirish** [comment]
> So from these snippets it sounds like Content-Type may be the stronger method.

contenttype is provided in the response headers.
mimeType is provided by the browser and is based on contentType.

since the browser can choose a mimetype even if a contenttype isn't provided, we should just go with mimetype. it's the signal that's more in-tune with how the browser read the content.

let's keep this check simple: if the mimeType isn't `text/html` we have a problem.

that's it. :)

**Comment by lemcardenas** [comment]
> let's keep this check simple: if the mimeType isn't `text/html` we have a problem.
> 
> that's it. :)

Thanks for the explanation! Sounds good to me :)

File: lighthouse-core/gather/gather-runner.js [hunk 13]
```diff
@@ -216,6 +216,28 @@ class GatherRunner {
...
+  static getNonHtmlError(mainRecord) {
+    // MIME types are case-insenstive
+    const HTML_MIME_REGEX = /^text\/html$/i;
+
+    // If we never requested a document, there's no doctype error, let other cases handle it.
+    if (!mainRecord) return undefined;
+
+    // mimeType is determined by the browser, we assume Chrome is determining mimeType correctly,
+    // independently of 'Content-Type' response headers, and always sending mimeType if well-formed.
+    if (mainRecord.mimeType || mainRecord.mimeType === '') {
```
Comment (connorjclark) [inline lighthouse-core/gather/gather-runner.js:233]: `mainRecord.mimeType` is initialized as an empty string and set from `response.mimeType` which apparently is always a string, never null or undefined. So I think this can be refactored to just:

```
if (!HTML_MIME_REGEX.test(mainRecord.mimeType)) {
  return new LHError(LHError.errors.NON_HTML, {mimeType: mainRecord.mimeType});
}
```

(removing the outer if statement)

File: lighthouse-core/lib/lh-error.js [hunk 14]
```diff
@@ -47,6 +47,11 @@ const UIStrings = {
   internalChromeError: 'An internal Chrome error occurred. Please restart Chrome and try re-running Lighthouse.',
   /** Error message explaining that fetching the resources of the webpage has taken longer than the maximum time. */
   requestContentTimeout: 'Fetching resource content has exceeded the allotted time',
+  /**
+   * @description Error message explaining that the webpage is non-HTML, so audits are ill-defined.
+   * @example {application/xml} mimeType
+   * */
+  nonHtml: 'The page provided is not HTML (served as MIME type {mimeType}).',
```
Comment (connorjclark) [inline lighthouse-core/lib/lh-error.js:54]: nit: `notHtml` instead? and `NOT_HTML`

File: lighthouse-core/gather/gather-runner.js [hunk 15]
```diff
@@ -216,6 +216,28 @@ class GatherRunner {
     return new LHError(LHError.errors.CHROME_INTERSTITIAL_ERROR);
   }
 
+  /**
+   * Returns an error if we try to load a non-HTML page.
+   * @param {LH.Artifacts.NetworkRequest|undefined} mainRecord
+   * @return {LH.LighthouseError|undefined}
+   */
+  static getNonHtmlError(mainRecord) {
+    // MIME types are case-insenstive
```
Comment (connorjclark) [inline lighthouse-core/gather/gather-runner.js:225]: there's no test for the case insensitivity

does chrome perhaps normalize this anyhow? you can hack the `static-server.js` to return any mime type you want and run `yarn static-server` to verify.

File: lighthouse-core/gather/gather-runner.js (hunk 13 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:233]: thanks for clarifying, I was unsure if the string from the response could be overwritten in the browser with a non-string / null / undefined, I'll make sure to fix this!

File: lighthouse-core/lib/lh-error.js (hunk 14 above)
Comment (lemcardenas) [reply lighthouse-core/lib/lh-error.js:54]: I kept the nonHtml name scheme from the function but this makes sense since the function isn't really in context here, I'll change this!

File: lighthouse-core/gather/gather-runner.js (hunk 15 above)
Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:225]: yeah I couldn't find information online as to whether chrome automatically makes mimetypes lowercase regardless of the http response, i'll check with the method you mentioned and let you know

Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:225]: ok update so I replaced all instances of 'text/html' with 'TEXT/HTML' in the static-server Content-Type header responses & confirmed in devtools that the Content-Type header showed up as 'TEXT/HTML'. When I ran a lighthouse run against a page hosted on static-server, I got no error and the mimeType accessible from mainRecord was always lowercase (text/html), so it seems that when chrome automatically makes its mimeType lowercase! 

So I am going to remove the case insensitivity modifier on the RegEx, which also makes me wonder if it should instead just be a direct string comparison with a const 'text/html' since thats likely more efficient and seems sufficient based on the test above

Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js:225]: nice! yeah a direct string comparison would be nice here. to be clear, the performance was never an issue (it's negligible), just the desire to keep things as complex as necessary (aka as simple as possible).

Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js:225]: a comment that chrome normalizes the mime type would be good too

Comment (lemcardenas) [reply lighthouse-core/gather/gather-runner.js:225]: sounds good! i'll make sure to add the comment right now

File: proto/lighthouse-result.proto [hunk 16]
```diff
@@ -55,6 +55,8 @@ enum LighthouseError {
   DNS_FAILURE = 19;
   // A timeout in the initial connection to the debugger protocol.
   CRI_TIMEOUT = 20;
+  // The page requested was non-HTML.
```
Comment (connorjclark) [inline proto/lighthouse-result.proto:58]:
```suggestion
  // The page requested was not HTML. 
```

Comment (lemcardenas) [reply proto/lighthouse-result.proto:58]: agh yeah I'll change that, I'll make sure to be consistent next time
